File: 463_Island_Perimeter/463_1.py
from typing import List

class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 1
            if grid[i][j] == 0:
                return 1
            if grid[i][j] == 2:
                return 0
            
            grid[i][j] = 2
            return dfs(i - 1, j) + dfs(i + 1, j) + dfs(i, j - 1) + dfs(i, j + 1)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    return dfs(i, j)
        
        return 0

File: 463_Island_Perimeter/463_1.c
int islandPerimeter(int** grid, int gridSize, int* gridColSize){
    int m = gridSize, n = *gridColSize;
    int res = 0;
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (grid[i][j] == 1) {
                if (i - 1 < 0  || (i - 1 >= 0 && grid[i - 1][j] != 1)) res++;
                if (i + 1 >= m || (i + 1 < m  && grid[i + 1][j] != 1)) res++;
                if (j - 1 < 0  || (j - 1 >= 0 && grid[i][j - 1] != 1)) res++;
                if (j + 1 >= n || (j + 1 < n  && grid[i][j + 1] != 1)) res++;
            }
        }
    }
    return res;
}
